contracts/StreamManager.vy

```diff
@@ -17,243 +16,313 @@
 
-num_streams: public(HashMap[address, uint256])
-streams: public(HashMap[address, HashMap[uint256, Stream]])
 
-
-owner: public(address)
 token_is_accepted: public(HashMap[IERC20, bool])
 
+# Global index of Streams
+num_streams: public(uint256)
+streams: public(HashMap[uint256, Stream])
+
+# Service Provider (has all Capabilities, also beneficiary of funding)
+controller: public(address)
+new_controller: public(address)
+new_controller_proposed: public(uint256)
+CONTROLLER_ACCEPTANCE_DELAY: constant(uint256) = 7 * 24 * 60 * 60  # 1 week
+
+event NewControllerProposed:
+    old: indexed(address)
+    new: indexed(address)
+
+
+event NewControllerAccepted:
+    old: indexed(address)
+    new: indexed(address)
+
+
+# Delegated Abilities
+flag Ability:
+    MODFIY_TOKENS
+    MODFIY_VALIDATORS
+    MODFIY_ACCESS
+    CANCEL_STREAMS
+
+capabilities: public(HashMap[address, Ability])
+
 
 event StreamCreated:
+    id: indexed(uint256)
+    owner: indexed(address)
     token: indexed(IERC20)
-    creator: indexed(address)
-    stream_id: indexed(uint256)
     amount_per_second: uint256
     start_time: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+    products: DynArray[bytes32, MAX_PRODUCTS]
+
+
+event StreamOwnershipUpdated:
+    id: indexed(uint256)
+    old: indexed(address)
+    new: indexed(address)
 
 
 event StreamFunded:
-    creator: indexed(address)
-    stream_id: indexed(uint256)
-    amount_added: uint256
+    id: indexed(uint256)
+    funder: indexed(address)
+    added: uint256
 
 
-event StreamCancelled:
-    creator: indexed(address)
-    stream_id: indexed(uint256)
-    amount_locked: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+event StreamClaimed:
+    id: indexed(uint256)
+    claimer: indexed(address)
+    exhausted: indexed(bool)
+    claimed: uint256
 
 
-event Claimed:
-    creator: indexed(address)
-    stream_id: indexed(uint256)
-    stream_exhausted: indexed(bool)
-    claimed_amount: uint256
+event StreamCancelled:
+    id: indexed(uint256)
+    canceller: indexed(address)
+    reason: indexed(bytes32)
+    refunded: uint256
 
 
 @deploy
 def __init__(
-    owner: address,
+    controller: address,
     min_stream_life: uint256,  # timedelta in seconds
-    validators: DynArray[Validator, MAX_VALIDATORS],
     accepted_tokens: DynArray[IERC20, 20],
+    validators: DynArray[Validator, MAX_VALIDATORS],
 ):
-    self.owner = owner
+    self.controller = controller
     MIN_STREAM_LIFE = min_stream_life
-    self.validators = validators
 
     for token: IERC20 in accepted_tokens:
         self.token_is_accepted[token] = True
 
+    self.validators = validators
+
 
 @external
-def set_validators(validators: DynArray[Validator, MAX_VALIDATORS]):
-    assert msg.sender == self.owner
-    self.validators = validators
+def transfer_control(new_controller: address):
+    # NOTE: can revoke transfer at any time calling this method with `self.controller`
+    assert msg.sender == self.controller
```

Comment:
Should we add a dev-message here? (and other similar spots)

Reply:
really just one spot, all the rest rely on the capability fallback onto "is this the controller?" and hence this is different

Reply:
080ef47